### PR TITLE
New signals

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
cc @gkoz 
cc @EPashkin

### CODE REVIEW

**Comment by EPashkin** [comment]
+1

File: src/signal.rs [hunk 1]
```diff
@@ -2305,3 +2305,224 @@ mod cell_renderer {
...
+    unsafe extern "C" fn insert_trampoline<T>(this: *mut GtkEditable,
+                                              new_text: *mut c_char,
+                                              new_text_length: c_int,
+                                              position: *mut c_uint,
+                                              f: &Box<Fn(&T, &str, Option<&mut u32>) + 'static>)
+    where T: IsA<Editable> {
+        callback_guard!();
+        if new_text_length != -1 {
+            sub_func(this,
+                     Vec::from_raw_parts(new_text as *mut c_uchar,
```
Comment (gkoz) [inline src/signal.rs]: You probably want `slice::from_raw_parts` here.

Comment (GuillaumeGomez) [inline src/signal.rs]: Arf! Silly me... That would have save me a lot of time! :-/

File: src/signal.rs [hunk 2]
```diff
@@ -2305,3 +2305,224 @@ mod cell_renderer {
...
+                                         new_text_length as usize,
+                                         new_text_length as usize).as_slice(),
+                     position,
+                     f);
+        } else {
+            sub_func(this, CStr::from_ptr(new_text).to_bytes(), position, f);
+        }
+    }
+
+    unsafe fn sub_func<T>(this: *mut GtkEditable,
```
Comment (gkoz) [inline src/signal.rs]: Why split the function in two?

Comment (GuillaumeGomez) [inline src/signal.rs]: It was for a borrow issue. It'll be solved when I'll replace `Vec::from_raw_parts` by `slice::from_raw_parts`.

File: src/signal.rs [hunk 3]
```diff
@@ -2305,3 +2305,224 @@ mod cell_renderer {
...
+                          buf: &[u8],
+                          position: *mut c_uint,
+                          f: &Box<Fn(&T, &str, Option<&mut u32>) + 'static>)
+    where T: IsA<Editable> {
+        let string = str::from_utf8(buf).unwrap();
+        if position.is_null() {
```
Comment (gkoz) [inline src/signal.rs]: Is there a reason to think it can be null?

Comment (GuillaumeGomez) [inline src/signal.rs]: Yes, it's written in the doc. :)

File: src/signal.rs [hunk 4]
```diff
@@ -2305,3 +2305,224 @@ mod cell_renderer {
...
     }
 }
+
+pub trait EditableSignals {
+    fn connect_changed<F>(&self, changed_func: F) -> u64
+        where F: Fn(&Self) + 'static;
+    fn connect_delete_text<F>(&self, delete_text_func: F) -> u64
+        where F: Fn(&Self, i32, i32) + 'static;
+    fn connect_insert_text<F>(&self, insert_text_func: F) -> u64
+        where F: Fn(&Self, &str, &mut u32) + 'static;
```
Comment (gkoz) [inline src/signal.rs]: We should probably pick either `u32` or `i32` for both signals.

File: src/signal.rs (hunk 3 above)
Comment (gkoz) [inline src/signal.rs]: Can't find it in the docs so far.

File: src/signal.rs (hunk 4 above)
Comment (GuillaumeGomez) [inline src/signal.rs]: Then it'll be `i32`.

File: src/signal.rs (hunk 3 above)
Comment (GuillaumeGomez) [inline src/signal.rs]: I can't find it again either. I removed it.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by EPashkin** [comment]
All `EntryCompletion` signals seems generated in #315, can you compare it and may be remove?

**Comment by EPashkin** [comment]
Or I need add ignore for it

**Comment by GuillaumeGomez** [comment]
I think it's best for me to remove them. I'll just compare with the generated code before hand.

**Comment by GuillaumeGomez** [comment]
@EPashkin: I checked, everything seems good so I remove it from this PR. Well done! :D

**Comment by EPashkin** [comment]
Forgot `trait EntryCompletionSignals`

**Comment by GuillaumeGomez** [comment]
Oh indeed. Fixed.

**Comment by EPashkin** [comment]
+1

**Comment by EPashkin** [comment]
I forgot disable `Editable` in `Gir.toml` :(
@GuillaumeGomez can you do it in this PR?

**Comment by GuillaumeGomez** [comment]
@EPashkin: Sure.

**Comment by EPashkin** [comment]
By way, we need redefine all 3 signals for `Editable` or can have mix: 2 auto + 1 manual?
IMHO only problems can arise with docs if any.

**Comment by GuillaumeGomez** [comment]
@EPashkin: Ok for you?

File: Gir.toml [hunk 5]
```diff
@@ -374,6 +374,19 @@ status = "generate"
         nullable = false
 
 [[object]]
+name = "Gtk.Editable"
+status = "generate"
+    [[object.signal]]
+    name = "changed"
```
Comment (EPashkin) [inline Gir.toml]: With `pattern = ".+"` we need only one `[[object.signal]]`

**Comment by GuillaumeGomez** [comment]
@EPashkin: Hum... I guess it's better to avoid to spread signals implementation everywhere.

**Comment by EPashkin** [comment]
Then add Regen and +1 from me.

File: Gir.toml (hunk 5 above)
Comment (GuillaumeGomez) [inline Gir.toml]: Nice trick. Done.

**Comment by EPashkin** [comment]
And remove `"Gtk.Editable",` in `generate=['

**Comment by EPashkin** [comment]
Gir don't allow double-spending :wink:

**Comment by GuillaumeGomez** [comment]
Sorry for the troubles. ^^'

Updated.

**Comment by EPashkin** [comment]
Still no Regen commit.

**Comment by GuillaumeGomez** [comment]
And done!

**Comment by EPashkin** [comment]
Thanks.

**Comment by GuillaumeGomez** [comment]
cc @gkoz

**Comment by gkoz** [comment]
Do you think you could include a gir update here to fix the gesture build error?

**Comment by GuillaumeGomez** [comment]
I'll do it tomorrow.

**Comment by GuillaumeGomez** [comment]
@gkoz: Done.

**Comment by EPashkin** [comment]
Thanks

**Comment by EPashkin** [comment]
@GuillaumeGomez squash Regens pls

**Comment by GuillaumeGomez** [comment]
Done.

**Comment by EPashkin** [comment]
Sorry, I missed that no "Gir update" commit. Can you add it before Regen?

**Comment by GuillaumeGomez** [comment]
What "no gir update"?

**Comment by EPashkin** [comment]
"Update gir" commit that changes sub-module "gir" version: must be b798f4f to equals to Regen.

**Comment by GuillaumeGomez** [comment]
I update gir module but didn't switch the two commits. Do you REALLY want this as well?

**Comment by EPashkin** [comment]
I prefer update auto in these steps: update gir (if needed), update `gir.toml`, do some not auto work, do single "Regen",
but if you don't want it, I do not insist

**Comment by GuillaumeGomez** [comment]
This PR has already been updated a lot of times and it'd force me to rework the two last commits. So if possible, I'd prefer to avoid doing it.

**Comment by EPashkin** [comment]
Just to reminder: it part of https://github.com/gtk-rs/gir/issues/234

**Comment by EPashkin** [comment]
@GuillaumeGomez Found another forgotten bug:
`transmute(gl_area_trampoline_res)` without `as usize`
Don't understand why I don't see it before.

**Comment by EPashkin** [comment]
GlArea will be removed after https://github.com/gtk-rs/gir/issues/239, so can be ignored for now.

**Comment by GuillaumeGomez** [comment]
So... I remove it or not?

**Comment by EPashkin** [comment]
IMHO better remove to remove annoying warning.

**Comment by GuillaumeGomez** [comment]
Hum... What warning? I don't see any from both travis and appveyor.

**Comment by EPashkin** [comment]
```
D:/eap/rust/0/gtk\src\signal.rs:1153:21: 1153:30 warning: `unsafe extern "C" fn(*mut ffi::GtkGLArea, i32, i32, &Box<for<'r> std::ops::Fn(&'r auto::gl_area::GLArea, i32, i32) + 'static>) {signal::gl_area::gl_area_trampoline_res}` is now zero-sized and has to be cast to a pointer before transmuting to `std::option::Option<unsafe extern "C" fn()>`, #[warn(transmute_from_fn_item_types)] on by default
D:/eap/rust/0/gtk\src\signal.rs:1153                     transmute(gl_area_trampoline_res), Box::into_raw(f) as *mut _)
                                                         ^~~~~~~~~
D:/eap/rust/0/gtk\src\signal.rs:1153:21: 1153:30 warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
D:/eap/rust/0/gtk\src\signal.rs:1153:21: 1153:30 note: for more information, see issue #19925 <https://github.com/rust-lang/rust/issues/19925>
   Compiling gtk-rs-examples v0.0.1 (file:///D:/eap/rust/0/examples)
```
It was appear when building examples with gtk_3_16,
It was main reason to mass adding 'as usize` but I missed this function,
I don't understand why I don't see it yesterday.

**Comment by GuillaumeGomez** [comment]
It's strange we didn't see earlier (well, I'm on gtk 3.10, so normal for me :p).

**Comment by GuillaumeGomez** [comment]
And done.

**Comment by EPashkin** [comment]
Thanks again.

**Comment by gkoz** [comment]
Perfect, thanks!

**Comment by EPashkin** [comment]
@GuillaumeGomez You planing adding manual SpinButtonSignals?

**Comment by GuillaumeGomez** [comment]
I plan to add all missing signals. So if it's in the list, then yes. Don't know when I'll however.

**Comment by EPashkin** [comment]
Its part of https://github.com/gtk-rs/gir/issues/234, so we don't plan generate it, so it must be in list ;)

**Pull request merged**